backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java

```diff
@@ -0,0 +1,188 @@
+   * object with the right scores.
+   */
+  @Test
+  public void checkRightScores() {
+    // add the attributes we chose to analyze for and scores to API & create API
+
+    // add each of the desiredTypes and a random float between 0 & 1 as their score 
+    for (AttributeType type: desiredTypes) {
+      Random rand = new Random();
+      
+      desiredScores.put(type, rand.nextFloat());
+    }
+
+    api = createMockAPI(desiredScores);
+    
+    // get the PerspectiveValues object and specifically the output analyses (scores from PerspectiveAPI)
+    PerspectiveValues values = PerspectiveServiceClient.analyze(api, desiredTypes, "foo");
+    Map<AttributeType, Float> outputScores = values.getAttributeTypesToScores();
+
+    // check all the right scores were returned from analyze()
+    for (AttributeType type: desiredTypes) {
+      Assert.assertEquals(desiredScores.get(type), outputScores.get(type), 0);
+    }
+  }
+
+  /** 
+   * Returns a mock version of the PerspectiveAPI which will return the scores
+   * passed in through desiredScores or which returns -1 when there is no desired score
+   * (i.e. the map does not contain that attribute type as a key).
+   *
+   * @param desiredScores a map of the scores you want the mockAPI to return 
+   *    (if you don't set any, the mock version of this API will always return -1)
+   * @return the mocked version of the PerspectiveAPI
+   */
+  private static PerspectiveAPI createMockAPI(Map<AttributeType, Float> desiredScores) {
+    // instantiate all the necessary mock objects
+    PerspectiveAPI mockAPI = mock(PerspectiveAPI.class);
+    AnalyzeCommentResponse mockResponse = mock(AnalyzeCommentResponse.class);
+    AttributeScore mockAttributeScore = mock(AttributeScore.class);
+    Score mockScore = mock(Score.class);
+
+    // when analyze is called for our mockAPI, return the mockResponse
+    when(mockAPI.analyze(any(AnalyzeCommentRequest.class))).thenReturn(mockResponse);
+
+    // when getAttributeScore is called we need to be able to grab the type it is looking for
+    // then, check if this AttributeType is in the map and if it is, grab its value, and
+    // set up the next few calls so that this value will be returned. if it isn't in the map,
+    // set up the next few calls so -1 will be returned (-1 signifies no particular value was desired).
+    when(mockResponse.getAttributeScore(any(AttributeType.class))).thenAnswer(new Answer() {
+      public Object answer(InvocationOnMock invocation) {
+        Object[] args = invocation.getArguments();
+        Object mock = invocation.getMock();
+
+        float fakeScore = -1;
+        AttributeType type = (AttributeType) args[0];
+
+        if (desiredScores.containsKey(type)) {
+          fakeScore = desiredScores.get(type);
+        }
+
+        when(mockAttributeScore.getSummaryScore()).thenReturn(mockScore);
+        when(mockScore.getValue()).thenReturn(fakeScore);
```

Comment:
Would it be safer to create a new instance of Score and AttributeScore each time?

My concern with restubbing a single instance is that if any code holds on to an instance, it would potentially see that instance's score value change over time.

Reply:
That makes sense. I'll just create a new instance every time.

Reply:
Done